nativelib/src/main/scala/scala/scalanative/runtime/dwarf/DWARF.scala

```diff
@@ -0,0 +1,726 @@
+    def is(tag: DWARF.Tag): Boolean =
+      abbrev.exists(_.tag == tag)
+
+    def getName: Option[UInt] = values.collectFirst {
+      case v if v._1.at == DWARF.Attribute.DW_AT_name =>
+        v._2.asInstanceOf[UInt]
+    }
+
+    def getLine: Option[Int] = values.collectFirst {
+      case v if v._1.at == DWARF.Attribute.DW_AT_decl_line =>
+        v._2 match {
+          case x: UShort => x.toInt
+          case x: UByte  => x.toInt
+          case _         => 0
+        }
+    }
+
+    def getLowPC: Option[Long] = values.collectFirst {
+      case v if v._1.at == DWARF.Attribute.DW_AT_low_pc =>
+        v._2.asInstanceOf[Long]
+    }
+
+    def getHighPC(lowPC: Long): Option[Long] =
+      // As of DWARF v4, DW_AT_high_pc can be a constant that represents the offset from low_pc
+      // > If the value of the DW_AT_high_pc is of class address, it
+      // > is the relocated address of the first location past the last instruction associated with the entity; if
+      // > it is of class constant, the value is an unsigned integer offset which when added to the low PC
+      // > gives the address of the first location past the last instruction associated with the entity.
+      // "DWARF Debugging Information Format Version 4" - 2.17.2 Contiguous Address Range
+      // https://dwarfstd.org/doc/DWARF4.pdf
+      values.collectFirst {
+        case v
+            if v._1.at == DWARF.Attribute.DW_AT_high_pc &&
+              DWARF.Form.isConstantClass(v._1.form) =>
+          val value = v._2.asInstanceOf[UInt]
+          lowPC + value.toLong
+        case v
+            if v._1.at == DWARF.Attribute.DW_AT_high_pc &&
+              DWARF.Form.isAddressClass(v._1.form) =>
+          v._2.asInstanceOf[Long]
+      }
+  }
+
+  case class Section(offset: UInt, size: Long)
+  case class Strings(buf: Array[Byte]) {
+    def read(at: UInt): String = {
+
+      // WARNING: lots of precision loss
+      assert(at < buf.length.toUInt)
+      val until = buf.indexWhere(_ == 0, at.toInt)
+
+      new String(buf.slice(at.toInt, until))
+    }
+  }
+  object Strings {
+    lazy val empty = Strings(Array.empty)
+    def parse(debug_str: Section)(implicit bf: BinaryFile): Strings = {
+      val pos = bf.position()
+      bf.seek(debug_str.offset.toLong)
+
+      val buf = Array.ofDim[Byte](debug_str.size.toInt)
+      bf.readFully(buf)
+      bf.seek(pos)
+
+      Strings(buf)
+    }
+  }
+
+  def parse(
+      debug_info: Section,
+      debug_abbrev: Section
+  )(implicit bf: BinaryFile): Vector[DIE] = {
+    bf.seek(debug_info.offset.toLong)
+    val end_offset = debug_info.offset.toLong + debug_info.size
+    def stop = bf.position() >= end_offset
+    val dies = Vector.newBuilder[DIE]
+    while (!stop) {
+      val die = DIE.parse(debug_info, debug_abbrev)
+      dies += die
+    }
+    dies.result()
+  }
+
+  object DIE {
+    private val abbrevCache = mutable.Map.empty[Long, Vector[Abbrev]]
```

Comment:
We cache the result of parsing `.debug_abbrev` section, without it, we parse the same `.debug_abbrev` section over and over again for the number of `CompileUnit` DIE.